index.d.ts

```diff
@@ -0,0 +1,293 @@
+  validate(predicate?: (fieldState: Field) => boolean): Promise<boolean>
+  serialize(): SerializedFields
+  setValues(fieldsPatch: SerializedFields): void
+  setErrors(errors: FieldErrors): void
+  submit(): Promise<void> /** @todo */
+}
+
+interface Event {
+  event: React.FormEvent<HTMLInputElement>
+  fieldProps: FieldState<string>
+  fields: Fields
+  form: Form
+}
+
+export interface BlurEvent extends Event {}
+export interface FocusEvent extends Event {}
+
+export interface ChangeEvent extends Event {
+  event: React.FormEvent<HTMLInputElement>
+  prevValue: string
+  nextValue: string
+}
+
+export type Rule = RegExp | ((args: ResolverArgs) => boolean)
+export type AsyncRule = RegExp | ((args: ResolverArgs) => AsyncRulePayload)
+
+export type BlurFunction = (args: BlurEvent) => void
+export type ChangeFunction = (args: ChangeEvent) => void
+export type FocusFunction = (args: FocusEvent) => void
+
+export interface FieldProps {
+  rule?: Rule
+  asyncRule?: AsyncRule
+  required?: boolean
+  skip?: boolean
+  onBlur?: BlurFunction
+  onChange?: ChangeFunction
+  onFocus?: FocusFunction
+}
+
+/** @todo Value generic */
+export interface FieldState<V> {
+  asyncRule?: AsyncRule
+  controlled: boolean
+  debounceValidate: () => any /** @todo */
+  errors: string[] | null
+  expected: boolean
+  fieldGroup?: string /** @todo */
+  fieldPath: string[]
+  focused: boolean
+  getRef: any /** @todo */
+  initialValue: string
+  invalid: boolean
+  mapValue: (value: V) => any
+  onBlur?: BlurFunction
+  onChange?: ChangeFunction
+  onFocus?: FocusFunction
+  pendingAsyncValidation?: boolean /** @todo */
+  reactiveProps: any /** @todo */
+  required: boolean
+  rule?: Rule
+  serialize: (value: V) => string
+  skip?: boolean
+  touched: boolean
+  type: string
+  valid: boolean
+  validAsync: boolean
+  validatedAsync: boolean
+  validatedSync: boolean
+  validating: boolean
+  validSync: boolean
+  value: V
+  valuePropName: string
+}
+
+export const fieldPresets: {
+  checkbox: CreateFieldOptions<boolean, 'checked'>
+  input: CreateFieldOptions<string>
+  radio: CreateFieldOptions<string>
+  select: CreateFieldOptions<string>
+  textarea: CreateFieldOptions<string>
+}
+
+export interface CreateFieldOptions<V, ValuePropName = 'value'> {
+  allowMultiple?: boolean
+  valuePropName?: ValuePropName
+  initialValue?: any
+  beforeRegister: () => any /** @todo */
+  shouldValidateOnMount: (
+    args: {
+      valuePropName: ValuePropName /** @todo [valuePropName] dynamic property */
+      fieldRecord: any
+      context: any
+    },
+  ) => boolean
+  mapPropsToField: () => any /** @todo */
+  enforceProps: () => Object /** @todo */
+  assertValue: (value: V) => boolean
+  mapValue: (value: any) => any
+  serialize: (value: V) => any
+}
+
+export function createField<P>(
+  options: CreateFieldOptions<string>,
+): (
+  component: (
+    props: P & {
+      fieldProps: React.InputHTMLAttributes<HTMLInputElement>
```

Comment:
`React.HTMLAttributes` might be a better choice, I thought about adding all variants but that'd require type assertion.